src/libcore/array.rs

```diff
@@ -264,6 +267,269 @@ macro_rules! array_impls {
+        where B: PartialEq<A>, [(); N]: AllowedArrayLength
+    {
+        #[inline]
+        fn eq(&self, other: &[A; N]) -> bool { self[..] == other[..] }
+        #[inline]
+        fn ne(&self, other: &[A; N]) -> bool { self[..] != other[..] }
+    }
+
+    #[stable(feature = "rust1", since = "1.0.0")]
+    impl<'a, 'b, A, B, const N: usize> PartialEq<&'b mut [B]> for [A; N]
+        where A: PartialEq<B>, [(); N]: AllowedArrayLength
+    {
+        #[inline]
+        fn eq(&self, other: &&'b mut [B]) -> bool { self[..] == other[..] }
+        #[inline]
+        fn ne(&self, other: &&'b mut [B]) -> bool { self[..] != other[..] }
+    }
+
+    #[stable(feature = "rust1", since = "1.0.0")]
+    impl<'a, 'b, A, B, const N: usize> PartialEq<[A; N]> for &'b mut [B]
+        where B: PartialEq<A>, [(); N]: AllowedArrayLength
+    {
+        #[inline]
+        fn eq(&self, other: &[A; N]) -> bool { self[..] == other[..] }
+        #[inline]
+        fn ne(&self, other: &[A; N]) -> bool { self[..] != other[..] }
+    }
+
+    // NOTE: some less important impls are omitted to reduce code bloat
+    // __impl_slice_eq2! { [A; $N], &'b [B; $N] }
+    // __impl_slice_eq2! { [A; $N], &'b mut [B; $N] }
+
+    #[stable(feature = "rust1", since = "1.0.0")]
+    impl<T:Eq, const N: usize> Eq for [T; N]
+        where [(); N]: AllowedArrayLength
+    { }
+
+    #[stable(feature = "rust1", since = "1.0.0")]
+    impl<T:PartialOrd, const N: usize> PartialOrd for [T; N]
+        where [(); N]: AllowedArrayLength
+    {
+        #[inline]
+        fn partial_cmp(&self, other: &[T; N]) -> Option<Ordering> {
+            PartialOrd::partial_cmp(&&self[..], &&other[..])
+        }
+        #[inline]
+        fn lt(&self, other: &[T; N]) -> bool {
+            PartialOrd::lt(&&self[..], &&other[..])
+        }
+        #[inline]
+        fn le(&self, other: &[T; N]) -> bool {
+            PartialOrd::le(&&self[..], &&other[..])
+        }
+        #[inline]
+        fn ge(&self, other: &[T; N]) -> bool {
+            PartialOrd::ge(&&self[..], &&other[..])
+        }
+        #[inline]
+        fn gt(&self, other: &[T; N]) -> bool {
+            PartialOrd::gt(&&self[..], &&other[..])
+        }
+    }
+
+    #[stable(feature = "rust1", since = "1.0.0")]
+    impl<T:Ord, const N: usize> Ord for [T; N]
+        where [(); N]: AllowedArrayLength
+    {
+        #[inline]
+        fn cmp(&self, other: &[T; N]) -> Ordering {
+            Ord::cmp(&&self[..], &&other[..])
+        }
+    }
+}
+
+/// Implemented for lengths where trait impls are allowed on arrays in core/std
+#[rustc_on_unimplemented(
+    label="trait not implemented for an array this long",
+    message="arrays only have std trait implementations for lengths 0..=32",
+)]
+#[unstable(feature = "allowed_array_length", issue = "0",
+    reason = "will never be stable, just a temporary step until const generics are stable")]
+#[cfg(not(bootstrap))]
+pub trait AllowedArrayLength {}
```

Comment:
Can you please add some UI tests (e.g. in `ui/const-generics/array-impls/`) for some of the simpler / more complex const generic impls above to ensure that direct (e.g. asserting the bound on a too big array type) and indirect (e.g. calling a method on a too big array) uses of these impls do not work?

Reply:
Added some in 4b0ad55; let me know whether that's what you were thinking.